2020/day-4-passport-processing/index.ts

```diff
@@ -0,0 +1,75 @@
+// byr (Birth Year) - four digits; at least 1920 and at most 2002.
+// iyr (Issue Year) - four digits; at least 2010 and at most 2020.
+// eyr (Expiration Year) - four digits; at least 2020 and at most 2030.
+// hgt (Height) - a number followed by either cm or in:
+//     If cm, the number must be at least 150 and at most 193.
+//     If in, the number must be at least 59 and at most 76.
+// hcl (Hair Color) - a # followed by exactly six characters 0-9 or a-f.
+// ecl (Eye Color) - exactly one of: amb blu brn gry grn hzl oth.
+// pid (Passport ID) - a nine-digit number, including leading zeroes.
+// cid (Country ID) - ignored, missing or not.
+
+const parseList = (list: string) => {
+  return list
+    .trim()
+    .split(/\n\n/)
+    .map(entry => entry.split(/\s+/).map(x => x.split(':')))
+    .map(Object.fromEntries)
+}
+
+type Passport = {
+  byr?: string
+  iyr?: string
+  eyr?: string
+  hgt?: string
+  hcl?: string
+  ecl?: string
+  pid?: string
+  cid?: string
+}
+
+type Maybe<T> = T | null | undefined
```

Comment:
Could you explain this syntax for me?

Reply:
This is a [TS "generic"](https://www.typescriptlang.org/docs/handbook/generics.html#generic-types). The `<T>` here acts basically like a function parameter, so this `Maybe` type is a small utility type which takes in any type you pass it and returns a type which is either the type you provided or `null` or `undefined`.

So basically the result of `Maybe<string>` is `string | null | undefined`